### PR TITLE
feat(notification): add cli preview for func hosted bot

### PR DESCRIPTION
**Pull request by a1exwang**
- add cli preview for func hosted bot
- fix a bug in backend preview (backendStart should be run after backendWatch)

The start of 'bot start' is after the end of 'bot watch'
image

* Update 1: change `sequentialTasks` to use callback to correctly handle sequentially run.

https://msazure.visualstudio.com/Microsoft%20Teams%20Extensibility/_workitems/edit/13776527/

### CODE REVIEW

**Comment by a1exwang** [comment]
E2E TEST: n/a

File: packages/cli/src/cmds/preview/preview.ts [hunk 1]
```diff
@@ -778,12 +779,132 @@ export default class Preview extends YargsCommand {
...
+        : this.prepareTask(
+            TaskDefinition.botStart(workspaceFolder, programmingLanguage, true),
+            constants.botStartStartMessage,
+            commonUtils.getBotLocalEnv(localEnv)
+          );
+
+      const botAzuriteTask = this.prepareTask(
+        TaskDefinition.funcHostedBotAzurite(workspaceFolder),
+        constants.botWatchStartMessage,
+        commonUtils.getBotLocalEnv(localEnv)
```
Comment (swatDong) [inline packages/cli/src/cmds/preview/preview.ts:866]: Azurite does not need bot env.

File: packages/cli/src/cmds/preview/preview.ts [hunk 2]
```diff
@@ -778,12 +779,132 @@ export default class Preview extends YargsCommand {
...
+                TaskDefinition.funcHostedBotWatch(workspaceFolder),
+                constants.botWatchStartMessage,
+                commonUtils.getBotLocalEnv(localEnv)
+              )
+          : undefined;
+      const botStartTask = includeFuncHostedBot
+        ? // For func hosted bot, always use the new task (prepareTaskNext).
+          this.prepareTaskNext(
+            TaskDefinition.funcHostedBotStart(workspaceFolder),
+            constants.botStartStartMessage,
```
Comment (swatDong) [inline packages/cli/src/cmds/preview/preview.ts:847]: botStartStartMessageNext

File: packages/fx-core/src/common/local/taskDefinition.ts [hunk 3]
```diff
@@ -103,6 +103,45 @@ export class TaskDefinition {
...
 
+  static funcHostedBotWatch(workspaceFolder: string): ITaskDefinition {
+    return {
+      name: "bot watch",
+      // command is never used for the new local preview in CLI.
+      command: "npm run watch:teamsfx",
+      cwd: path.join(workspaceFolder, FolderName.Bot),
+      execOptions: {
+        needShell: true,
+        needCmd: isWindows(),
```
Comment (kuojianlu) [inline packages/fx-core/src/common/local/taskDefinition.ts:114]: curious why `needCmd` is required?

File: packages/fx-core/src/common/local/taskDefinition.ts [hunk 4]
```diff
@@ -103,6 +103,45 @@ export class TaskDefinition {
...
+  }
+
+  static funcHostedBotStart(workspaceFolder: string): ITaskDefinition {
+    return {
+      name: "bot start",
+      // command is never used for the new local preview in CLI.
+      command: `npm run dev:teamsfx`,
+      cwd: path.join(workspaceFolder, FolderName.Bot),
+      execOptions: {
+        needShell: true,
```
Comment (kuojianlu) [inline packages/fx-core/src/common/local/taskDefinition.ts:127]: `backendStart` has `needCmd`, what about `funcHostedBotStart`?

File: packages/cli/src/cmds/preview/constants.ts [hunk 5]
```diff
@@ -35,10 +35,17 @@ export const remoteTeamsAppIdConfigKey = "teamsAppId";
...
+// From vscode $tsc-watch problem matcher: https://github.com/microsoft/vscode/blob/5a0ab56492d0c99f08028ca62ac3d59edb37f30f/extensions/typescript-language-features/package.json#L1085
+export const tscWatchPattern =
+  /^\s*(?:message TS6042:|\[?\D*\d{1,2}[:.]\d{1,2}[:.]\d{1,2}\D*(├\D*\d{1,2}\D+┤)?(?:\]| -)) (?:Compilation complete\.|Found \d+ errors?\.) Watching for file changes\./g;
+// make a copy to prevent accidental change
+export const backendWatchPattern = new RegExp(tscWatchPattern);
+export const funcHostedBotWatchPattern = new RegExp(tscWatchPattern);
 export const authStartPattern = /.*/g;
 export const ngrokStartPattern = /started tunnel|failed to reconnect session/g;
 export const botStartPattern = /listening|[nodemon] app crashed/g;
+export const funcHostedBotStartPattern = /Compiled|Failed/g;
```
Comment (kuojianlu) [inline packages/cli/src/cmds/preview/constants.ts:47]: Is `/Compiled|Failed/g` correct?

File: packages/fx-core/src/common/local/taskDefinition.ts (hunk 3 above)
Comment (a1exwang) [reply packages/fx-core/src/common/local/taskDefinition.ts:114]: copied from backend start
image

File: packages/cli/src/cmds/preview/constants.ts (hunk 5 above)
Comment (a1exwang) [reply packages/cli/src/cmds/preview/constants.ts]: fixed

File: packages/fx-core/src/common/local/taskDefinition.ts (hunk 3 above)
Comment (a1exwang) [reply packages/fx-core/src/common/local/taskDefinition.ts:114]: Copied from bot start and bot watch task:
image

File: packages/fx-core/src/common/local/taskDefinition.ts (hunk 4 above)
Comment (a1exwang) [reply packages/fx-core/src/common/local/taskDefinition.ts:127]: see the last comment

File: packages/cli/src/cmds/preview/preview.ts (hunk 1 above)
Comment (a1exwang) [reply packages/cli/src/cmds/preview/preview.ts]: fixed

File: packages/cli/src/cmds/preview/preview.ts (hunk 2 above)
Comment (a1exwang) [reply packages/cli/src/cmds/preview/preview.ts]: fixed

File: packages/fx-core/src/common/local/taskDefinition.ts (hunk 4 above)
Comment (kuojianlu) [reply packages/fx-core/src/common/local/taskDefinition.ts:127]: So the start task requires `needCmd`, but watch task does not.

**Review by swatDong — DISMISSED** [review]

**Review by Alive-Fish — APPROVED** [review]

**Review by swatDong — APPROVED** [review]

**Pull request merged**